File: java/challenges/NthElementFromTheEnd.java
public final class NthElementFromTheEnd {

//Singly-linked lists are already defined with this interface:
// class ListNode<T> {
//     ListNode(T x) {
//       value = x;
//     }
//     T value;
//     ListNode<T> next;
// }
int nthElementFromTheEnd(ListNode<Integer> l, int n) {
        ArrayList<Integer> out = new ArrayList();
        ListNode<Integer> ln = l;
        while(ln != null) {
            out.add(ln.value);
            ln = ln.next;
        }
        return (n > out.size()) ? -1 : out.get(out.size() - n);
    }
}

File: ruby/challenges/nthElementFromTheEnd.rb
# Singly-linked lists are already defined with this interface:
# class ListNode
#   attr_accessor :value, :next
#   def initialize(val)
#     @value = val
#     @next = nil
# end
#
def nthElementFromTheEnd(l, n)
    out, ln = [], l
    while ln do
        out.push ln.value
        ln = ln.next
    end
    return n > out.size ? -1 : out[out.size - n]
end
